### PR TITLE
fs-1524-mark-qa-complete

### PR DESCRIPTION
**Pull request by gio-karunakaran**

### CODE REVIEW

**Comment by gio-karunakaran** [comment]
closing PR as easier to re-start

File: app/assess/routes.py [hunk 1]
```diff
@@ -219,6 +224,38 @@ def flag(application_id):
     )
 
 
+@assess_bp.route("/qa_complete/<application_id>", methods=["GET", "POST"])
+@login_required(roles_required=["LEAD_ASSESSOR"])
+def qa_complete(application_id):
+
+    form = MarkQaCompleteForm()
+
+    if request.method == "POST" and form.validate_on_submit():
```
Comment (tferns) [inline app/assess/routes.py:233]: `form.validate_on_submit()` already checks for `POST` method: https://github.com/wtforms/flask-wtf/blob/1cf8b19bcbc53e5357a4aa0c5ccb5cf7d3ea0d22/src/flask_wtf/form.py#L127

**Review by tferns — CHANGES_REQUESTED** [review]
Looks good but can we add some test coverage for these changes?

File: app/assess/templates/assessor_tasklist.html [hunk 2]
```diff
@@ -5,9 +5,12 @@
 {% from 'macros/criteria_element.html' import criteria_element -%}
```
Comment (tferns) [inline app/assess/templates/assessor_tasklist.html:5]: This file in particular has a lot of conditional logic in the Jinja templates, I think we'd be more confident in this logic if we had some test coverage, would also be easier to develop on in the future I think

Comment (gio-karunakaran) [reply app/assess/templates/assessor_tasklist.html:5]: great point, added tests 👍🏾

File: app/assess/routes.py (hunk 1 above)
Comment (gio-karunakaran) [reply app/assess/routes.py]: done!

File: app/assess/routes.py [hunk 3]
```diff
@@ -217,6 +222,38 @@ def flag(application_id):
     )
 
 
+@assess_bp.route("/qa_complete/<application_id>", methods=["GET", "POST"])
+@login_required(roles_required=["LEAD_ASSESSOR"])
+def qa_complete(application_id):
+
```
Comment (ramsharma-prog) [inline app/assess/routes.py:228]: Add some docs please

Comment (gio-karunakaran) [reply app/assess/routes.py:234]: done!

File: tests/test_routes.py [hunk 4]
```diff
@@ -497,6 +497,92 @@ def test_post_continue_application(self, flask_test_client, mocker):
...
+        )
+        app.assess.routes.get_latest_flag.assert_called_once()
+        app.assess.routes.get_latest_flag.assert_called_once_with("app_123")
+
+        assert response.status_code == 200
+        assert b"Marked as QA complete" in response.data
+        assert b"01/01/2023 at 00:00am" in response.data
+        assert b"Flagged" not in response.data
+
+    def test_qa_completed_flagged_application(self, flask_test_client, mocker):
```
Comment (ramsharma-prog) [inline tests/test_routes.py:543]: Like the test ❤️

File: tests/test_routes.py [hunk 5]
```diff
@@ -497,6 +497,92 @@ def test_post_continue_application(self, flask_test_client, mocker):
 
         assert response.status_code == 302
         assert response.headers["Location"] == "/assess/application/app_123"
+
+    def test_qa_complete_flag_displayed(self, flask_test_client, mocker):
+        token = create_valid_token(test_lead_assessor_claims)
+        flask_test_client.set_cookie("localhost", "fsd_user_token", token)
+        mocker.patch(
```
Comment (Adam-W1) [inline tests/test_routes.py:504]: Im just conscious that the data we're using is diverging across PRs, some tests use test data in endpoint_json.json file and some a manual mock. I think the cleaner approach is probably using the .json file as we can also then see this when using flask run locally. Happy to discuss

File: config/envs/development.py [hunk 6]
```diff
@@ -3,7 +3,6 @@
 from os import getenv
 
 from config.envs.default import DefaultConfig
-from fsd_utils import CommonConfig
```
Comment (Adam-W1) [inline config/envs/development.py:6]: This needs re-adding

Comment (gio-karunakaran) [reply config/envs/development.py:6]: we're not using this anywhere in the file?

File: app/assess/templates/assessor_tasklist.html [hunk 7]
```diff
@@ -5,39 +5,44 @@
...
-        <div class="govuk-grid-column-full">
-            <div class="assessment-complete  govuk-margin-bottom-8">
-                <h2 class="assessment-alert__heading govuk-heading-l govuk-!-margin-top-4">
-                    Assessment complete</h2>
-                <p class="govuk-body">You have marked this assessment as complete.</p>
-                <p class="govuk-body">
-                    It may now be selected for quality assurance (QA). New scores and rationales could be added if this
-                    happens.
-                </p>
+    {% elif state.workflow_status=="COMPLETED" and not flag.is_qa_complete %}
```
Comment (Adam-W1) [inline app/assess/templates/assessor_tasklist.html:32]: This refactored code block seems to not show assessment complete if there is a flag, I think this is contrary to seeing flags and the completion message on the same page as mentioned here https://communities-govuk.slack.com/archives/C04L4H0218U/p1674645287629009. *We should be able to run a flag workflow on a completed application.

File: tests/test_routes.py (hunk 5 above)
Comment (gio-karunakaran) [reply tests/test_routes.py:504]: yeh definitely worth a chat! could be worth raising a separate BAU ticket to make these tests consistent - although I much prefer using a mock 😆

File: config/envs/development.py (hunk 6 above)
Comment (Adam-W1) [reply config/envs/development.py:6]: lines 23-25 as part of the USE_LOCAL_DATA block i think

File: app/assess/templates/assessor_tasklist.html (hunk 7 above)
Comment (gio-karunakaran) [reply app/assess/templates/assessor_tasklist.html]: done!

File: tests/test_routes.py (hunk 5 above)
Comment (Adam-W1) [reply tests/test_routes.py:504]: yeah, Im happy o raise a BAU following our chat earlier. 
Test mocks and USE_LOCAL_DATA calls in endpoint_json.json for flask run, consuming the same data (local to test directory) is a good solution 👍

File: config/envs/development.py (hunk 6 above)
Comment (gio-karunakaran) [reply config/envs/development.py:6]: you're right, have added now 👍🏾

**Review by tferns — APPROVED** [review]

**Pull request merged**